src/common/Throttle.h

```diff
@@ -338,16 +338,58 @@ class TokenBucketThrottle {
   struct Bucket {
     CephContext *cct;
     const std::string name;
-    std::atomic<uint64_t> remain = { 0 }, max = { 0 };
 
-    Bucket(CephContext *cct, const std::string& n, uint64_t m)
-      : cct(cct), name(n),
-	remain(m), max(m)
-    {
-    }
+    // minimum of the filling period.
+    static const uint64_t tick_min = 50;
+
+    // CIR: the rate tokens generated
+    uint64_t committed_information_rate = 0;
+    // CBS: the size of the bucket(also means the burst limit size)
+    uint64_t committed_burst_size = 0;
+    // tokens remain in the bucket after get()
+    uint64_t remain = 0;
+
+    // tokens filling period, its unit is millisecond.
+    uint64_t tick = tick_min;
+
+    /**
+     * These variables are used to calculate how many tokens need to be filled
+     * within each tick.
+     *
+     * In actual use, the tokens per tick may not be an integer, so take the its
+     * integer part here(tokens_per_tick) and spread the remainder(divide_left)
+     * into each tick to ensure that a set number of tokens can be transmitted
+     * per second.
+     *
+     * For exmple, we set the value of CIR to be 950, means 950 iops(or bps).
+     * In this case, the actual period is 1000 / 950 = 1.052, too small for
+     * the SafeTimer. So the tick should set to be 50(tick_min), 20 ticks in
+     * one second.
+     * The tokens filled in bucket per tick is 950 / 20 = 47.5, not an integer.
+     * We take the 47 as tokens_per_tick, and there are 950 - (47 * 20) = 10
+     * tokens left. To fix this, we spread the 10 tokens into 10 ticks.
+     *
+     * As a result, the tokens filled in one second will shown as this:
+     * tick    | 1| 2| 3| 4| 5| 6| 7| 8| 9|10|11|12|13|14|15|16|17|18|19|20|
+     * tokens  |48|48|48|48|48|48|48|48|48|48|47|47|47|47|47|47|47|47|47|47|
+     */
+    uint64_t ticks_per_second = 0;
+    uint64_t tokens_per_tick = 0;
+    uint64_t divide_left = 0;
+
+    Bucket(CephContext *cct,
+           const std::string& name,
+           uint64_t average,
+           uint64_t burst)
+    : cct(cct), name(name),
+      committed_information_rate(average),
+      committed_burst_size(burst) {}
 
     uint64_t get(uint64_t c);
-    uint64_t put(uint64_t c);
+    uint64_t put();
+    // set the committed_information_rate
+    bool set_average(uint64_t c);
+    // set the committed_burst_size
```

Comment:
@iridescent-rsy  thanx for the detailed comment here, it's helpful for review. But there is a comment about the design. Class Bucket should just be a "bucket", it should not take so many works here. I mean, the ticks_per_second, tokens_per_tick, devide_left, set_argerage() should be in the TokenBucketThrottle. 

Bucket should only have a simple design, name, max, remain, get() put(). that's all. But how to use this bucket, as you wish in upper user, TokenBucketThrottle. There is only a user of Bucket, tokenBucketThrottle, but if we want a leakyBucketThrottle in future, we will reuse the bucket. 

So move the Throttling related information from Bucket to TokenBucketThrottle.

Reply:
@yangdongsheng Thanks for your advice. I'll fix it immediately.